src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+
+        private bool _disposed;
+
+        private DefaultPipeReader _reader;
+        private DefaultPipeWriter _writer;
+
+        internal long Length => _length;
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> using <see cref="PipeOptions.Default"/> as options.
+        /// </summary>
+        public Pipe() : this(PipeOptions.Default)
+        {
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> with the specified <see cref="PipeOptions"/>.
+        /// </summary>
+        public Pipe(PipeOptions options)
+        {
+            if (options == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.options);
+            }
+
+            _bufferSegmentPool = new BufferSegment[SegmentPoolSize];
+
+            _readingState = default;
+            _readerCompletion = default;
+            _writerCompletion = default;
+
+            _pool = options.Pool;
+            _minimumSegmentSize = options.MinimumSegmentSize;
+            _maximumSizeHigh = options.PauseWriterThreshold;
+            _maximumSizeLow = options.ResumeWriterThreshold;
+            _readerScheduler = options.ReaderScheduler ?? PipeScheduler.Inline;
+            _writerScheduler = options.WriterScheduler ?? PipeScheduler.Inline;
+            _readerAwaitable = new PipeAwaitable(completed: false);
+            _writerAwaitable = new PipeAwaitable(completed: true);
+            _reader = new DefaultPipeReader(this);
+            _writer = new DefaultPipeWriter(this);
+        }
+
+        private void ResetState()
+        {
+            _readerCompletion.Reset();
+            _writerCompletion.Reset();
+            _commitHeadIndex = 0;
+            _currentWriteLength = 0;
+            _length = 0;
+        }
+
+        /// <summary>
+        /// Allocates memory from the pipeline to write into.
+        /// </summary>
+        /// <param name="minimumSize">The minimum size buffer to allocate</param>
+        internal Memory<byte> GetMemory(int minimumSize)
+        {
+            if (_writerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (minimumSize < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumSize);
+            }
+
+            lock (_sync)
+            {
+                BufferSegment segment = _writingHead ?? AllocateWriteHeadUnsynchronized(minimumSize);
+
+                int bytesLeftInBuffer = segment.WritableBytes;
+
+                // If inadequate bytes left or if the segment is readonly
+                if (bytesLeftInBuffer == 0 || bytesLeftInBuffer < minimumSize || segment.ReadOnly)
+                {
+                    BufferSegment nextSegment = CreateSegmentUnsynchronized();
+
+                    nextSegment.SetMemory(_pool.Rent(Math.Max(_minimumSegmentSize, minimumSize)));
```

Comment:
Do you need to call rent under the lock? Since pool is an abstraction, we don't know what rent does and so it can cause deadlocks or other synchronization problems.

Reply:
We need to do everything around it in lock. It might deadlock if pool would try to access the same pipe object on different thread which is not very common situation.

Reply:
OK. Let's see if we can optimized the locks later.